core/slim_globals.cpp

```diff
@@ -1421,9 +1425,13 @@ R"V0G0N({"$schema":"http://json-schema.org/schema#","additionalProperties":false
 
+const std::string gSLiM_tsk_population_metadata_schema = 
+R"V0G0N({"$schema":"http://json-schema.org/schema#","codec":"json","description":"SLiM schema for population metadata.","examples":[{"bounds_x0":0.0,"bounds_x1":100.0,"bounds_y0":0.0,"bounds_y1":100.0,"female_cloning_fraction":0.25,"male_cloning_fraction":0.0,"migration_records":[{"migration_rate":0.9,"source_subpop":1},{"migration_rate":0.1,"source_subpop":2}],"selfing_fraction":0.5,"sex_ratio":0.5,"slim_id":2,"name":"p2"}],"properties":{"bounds_x0":{"description":"The minimum x-coordinate in this subpopulation.","type":"number"},"bounds_x1":{"description":"The maximum x-coordinate in this subpopulation.","type":"number"},"bounds_y0":{"description":"The minimum y-coordinate in this subpopulation.","type":"number"},"bounds_y1":{"description":"The maximum y-coordinate in this subpopulation.","type":"number"},"bounds_z0":{"description":"The minimum z-coordinate in this subpopulation.","type":"number"},"bounds_z1":{"description":"The maximum z-coordinate in this subpopulation.","type":"number"},"description":{"description":"A description of this subpopulation.","type":"string"}},"female_cloning_fraction":{"description":"The frequency with which females in this subpopulation reproduce clonally (for WF models).","type":"number"},"male_cloning_fraction":{"description":"The frequency with which males in this subpopulation reproduce clonally (for WF models).","type":"number"},"migration_records":{"items":{"properties":{"migration_rate":{"description":"The fraction of children in this subpopulation that are composed of 'migrants' from the source subpopulation (in WF models).","type":"number"},"source_subpop":{"description":"The ID of the subpopulation migrants come from (in WF models).","type":"integer"}},"required":["source_subpop","migration_rate"],"type":"object"},"type":"array"},"name":{"description":"A human-readable name for this subpopulation.","type":"string"},"selfing_fraction":{"description":"The frequency with which individuals in this subpopulation self (for WF models).","type":"number"},"sex_ratio":{"description":"This subpopulation's sex ratio (for WF models).","type":"number"},"slim_id":{"description":"The ID of this population in SLiM. Note that this is called a 'subpopulation' in SLiM.","type":"integer"},"required":["slim_id"],"type":["object","null"]})V0G0N";
```

Comment:
Do we need to make `slim_id` required? What if we just ignored populations without a `slim_id` key?

Reply:
Well, populations that have no metadata at all do get ignored on the assumption that they are "dummies"; see slim_sim.cpp:8296.  We could do that for pops with metadata but without a `slim_id` key, but that seems kinda weird to me; such pops are probably *not* "dummies", so it would probably be surprising to the user if the file read in without errors but those pops just... weren't there.  I guess we're requiring a top-level SLiM metadata entry with a file version it in, etc., so this would not happen if someone just tried to import a .trees directly from msprime into SLiM; they would still (I think?) get an error that the file was not in SLIM format.  Given that, it's probably kinda moot, unless there's a desire/need to have pyslim write out pops that have metadata but do not have a `slim_id`.  So if it's moot, I'd prefer to keep it the way it is here, since just ignoring pops that do not look like "dummies" but that we don't understand seems wrong – not what the user probably wants, at best, and a confusing cause of bugs, at worst.  Is there an argument in favor of it?

Reply:
Here's the motivation: suppose someone runs an msprime simulations with a bunch of pops, but all the samples are in a single one, say population 5. Then they want to annotate it for SLiM. They might think that they just need to add a `slim_id` to only population 5, because that's the only one SLiM will actually use. So, we'll have to tell them that in fact all populations need a slim_id. This is not a big deal, I guess, since they'll have to be annotating other things anyhow.

Reply:
Hmm, or they could remove the metadata from the other pops that they don't want to show up in SLiM; if there's no metadata, SLiM will then ignore those subpops.  They have a decision to make: do they want those other subpops to be visible in SLiM or not?  And the required key forces them to make that decision, one way or the other.  That seems like a good thing to me...?

Reply:
Removing metadata isn't a good option, since it removes information about those populations (which are still used in the tree sequence!).

Reply:
Hmm; but if they don't want those subpops to load in SLiM at all, then by definition they don't care about that information, right?  They're prepared to drop those subpops completely; given that, the metadata attached to them doesn't seem like it matters.  Either the user wants them – add a slim_id – or they don't – remove the metadata and they will be dropped.  Is there a problem with that perspective, really?

Reply:
TBC I don't think this is a big deal. But, for clarity...

Nodes in the tree sequence that aren't part of any extant individuals still have `population` attributes; these nodes will never be read by SLiM and so may refer to populations that SLiM will never think about. If we start SLiM off at 200 years ago, does SLiM need a notion of "Anatolia from 10-15Kya" if that's what one of the "populations" in the tree sequence refers to? We certainly wouldn't want to drop the information.

Reply:
But SLiM *will* drop the information, if we do as you propose.  That's why I'm puzzled.  There are only two choices: load the subpop or don't load the subpop.  To load the subpop, we need a `slim_id`.  If we're _not_ going to load the subpop, then it makes sense to me to make the user "state" that intention explicitly by removing the metadata for the subpop, designating it as, in effect, nonexistent; that metadata would not be used anyway, if the subpop was not loaded into SLiM, right?  So we haven't lost anything that would not have been lost anyway.  The worst of all possible worlds, it seems to me, is to silently drop subpopulations from the model (which, as I understand it, is what you are advocating), simply because they lack a `slim_id`; I can easily see that causing confusion and difficulty.  I feel like we're talking past each other, but I don't see the benefit of what you propose, and the drawbacks seem clear.  What am I missing?

Reply:
Ah, I see. No, sorry, I was imagining that entries in the population table that SLiM didn't use would be just left as they were. =)

Reply:
I'm not sure I know what you mean.  Perhaps I'm being dense.

Reply:
I meant that I was being dense, I wasn't thinking properly about how SLiM rewrites the population table.

Reply:
Ah!  OK, then we're done.  :->  Of course if this architecture proves to be a problem in future, we can revisit it.  Something to look forward to.  ;->